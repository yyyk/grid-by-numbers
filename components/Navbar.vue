<template>
  <div class="nav">
    <v-app-bar app absolute flat>
      <v-app-bar-nav-icon @click.stop="drawer = !drawer" />
      <v-toolbar-title>
        <nuxt-link to="/" exact class="text--primary header-title-link">Grid By Numbers</nuxt-link>
      </v-toolbar-title>
      <v-spacer />
      <v-btn
        v-if="showEditorLink"
        depressed
        nuxt
        exact
        tag="a"
        color="primary"
        to="/editor"
        class="editor-link"
        aria-label="Link to Code Editor"
      >editor</v-btn>
      <v-btn
        v-if="showCompileButton"
        depressed
        color="primary"
        class="compile-button"
        aria-label="Compile Code"
        @click="onCompile"
      >compile</v-btn>
    </v-app-bar>
    <v-navigation-drawer
      v-model="drawer"
      app
      :disable-route-watcher="true"
      :disable-resize-watcher="true"
      floating
    >
      <v-list nav dense :expand="true">
        <v-subheader class="mt-2">
          <span class="nav-drawer-header">Table of Contents</span>
        </v-subheader>
      </v-list>
      <v-list-item
        v-for="(link, index) in menu"
        :key="index"
        nuxt
        exact
        :to="link.path === '/index' ? '/' : link.path"
        color="primary"
        class="link-text"
      >
        <v-list-item-content>
          <v-list-item-title v-text="link.title" />
        </v-list-item-content>
      </v-list-item>
    </v-navigation-drawer>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapGetters } from 'vuex'

export default Vue.extend({
  data: () => ({
    drawer: false,
    showEditorLink: false,
    showCompileButton: false
  }),
  computed: {
    ...mapGetters(['menu'])
  },
  watch: {
    $route() {
      this.setButtonVisibility()
    }
  },
  created() {
    this.setButtonVisibility()
  },
  methods: {
    setButtonVisibility() {
      const fullPath = this.$route.fullPath
      if (fullPath && /^\/editor/i.test(fullPath)) {
        this.showEditorLink = false
        this.showCompileButton = true
      } else {
        this.showEditorLink = true
        this.showCompileButton = false
      }
    },
    onCompile() {
      this.$nuxt.$emit('compile')
    }
  }
})
</script>

<style lang="scss" scoped>
.header-title-link {
  text-decoration: none;
}

.link-text {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}
</style>
