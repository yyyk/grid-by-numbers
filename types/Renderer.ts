import { ProductObject } from '@/lib/gbnscript/src/product'

export interface GridData {
  id: string
  tag: string
  children: ProductObject[]
}
