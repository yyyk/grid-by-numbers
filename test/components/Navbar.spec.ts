import Vue from 'vue'
import Vuex from 'vuex'
import Vuetify from 'vuetify'
import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import Navbar from '@/components/Navbar.vue'
import { MenuLink } from '@/types/Menu'

Vue.use(Vuetify)

const localVue = createLocalVue()
localVue.use(Vuex)

const store = new Vuex.Store({
  state: {
    menu: [
      {
        title: 'title1',
        path: '/index'
      },
      {
        title: 'title2',
        path: '/path2'
      },
      {
        title: 'title3',
        path: '/path3'
      }
    ] as MenuLink[]
  },
  getters: {
    menu: (state) => state.menu
  }
})

describe('Navbar.vue', () => {
  let vuetify: any

  const mountFunction = (options = {}) => {
    return mount(Navbar, {
      store,
      localVue,
      vuetify,
      stubs: { NuxtLink: RouterLinkStub },
      mocks: {
        $route: {
          fullPath: ''
        }
      },
      ...options
    })
  }

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  it('renders correctly', () => {
    const wrapper = mountFunction()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('exists', () => {
    const wrapper = mountFunction()
    expect(wrapper.exists()).toBe(true)
  })

  it('should have nav css class', () => {
    const wrapper = mountFunction()
    const parentDiv = wrapper.find('.nav')
    expect(parentDiv.exists()).toBe(true)
  })

  it('should have a title link', () => {
    const wrapper = mountFunction()

    const title = wrapper.find('.header-title-link')
    expect(title.exists()).toBe(true)
  })

  describe('editor/compile button', () => {
    it('should show editor button if the fullPath is not /editor', async () => {
      const wrapper = mountFunction({
        mocks: {
          $route: {
            fullPath: '/'
          }
        }
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$data.showEditorLink).toBe(true)
      const button = wrapper.find('.editor-link')
      expect(button.exists()).toBe(true)
    })

    it('should show compile button if the fullPath is /editor', async () => {
      const wrapper = mountFunction({
        mocks: {
          $route: {
            fullPath: '/editor'
          }
        }
      })
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$data.showEditorLink).toBe(false)
      const button = wrapper.find('.compile-button')
      expect(button.exists()).toBe(true)
    })

    it("should emit 'compile' when compile button is clicked", async () => {
      const emit = jest.fn()
      const wrapper = mountFunction({
        mocks: {
          $route: {
            fullPath: '/editor'
          },
          $nuxt: {
            $emit: emit
          }
        }
      })
      await wrapper.vm.$nextTick()
      const button = wrapper.find('.compile-button')
      button.trigger('click')
      expect(emit).toHaveBeenCalledWith('compile')
    })
  })

  describe('Nav drawer', () => {
    it('should have drawer closed by default', () => {
      const wrapper = mountFunction()
      expect(wrapper.vm.$data.drawer).toBe(false)
    })

    it('should open when .v-app-bar__nav-icon is clicked', async () => {
      const wrapper = mountFunction()
      wrapper.setData({
        drawer: false
      })
      const btn = wrapper.find('.v-app-bar__nav-icon')
      await btn.trigger('click')
      expect(wrapper.vm.$data.drawer).toBe(true)
    })

    it('should have a title', () => {
      const wrapper = mountFunction()

      const title = wrapper.find('.v-navigation-drawer .nav-drawer-header')
      expect(title.exists()).toBe(true)
    })

    it('should have .link-text if menu is not empty', () => {
      const wrapper = mountFunction()
      const links = wrapper.findAll('.v-navigation-drawer .link-text')
      expect(links.length).toBeGreaterThan(0)
    })

    it("should have .link-text replace '/index' with '/'", () => {
      const wrapper = mountFunction()
      const link = wrapper.findAll('.v-navigation-drawer .link-text').at(0)
      expect(link.props('to')).toBe('/')
    })
  })
})
