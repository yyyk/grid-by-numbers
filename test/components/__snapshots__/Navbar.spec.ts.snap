// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar.vue renders correctly 1`] = `
"<div class=\\"nav\\">
  <header class=\\"v-sheet theme--light v-toolbar v-toolbar--absolute v-toolbar--flat v-app-bar\\" style=\\"height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;\\">
    <div class=\\"v-toolbar__content\\" style=\\"height: 64px;\\"><button type=\\"button\\" class=\\"v-app-bar__nav-icon v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-menu theme--light\\"></i></span></button>
      <div class=\\"v-toolbar__title\\"><a class=\\"text--primary header-title-link\\">
          Grid By Numbers
        </a></div>
    </div>
  </header>
  <nav class=\\"v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--floating v-navigation-drawer--is-mobile theme--light\\" style=\\"height: 100vh; top: 0px; transform: translateX(-100%); width: 256px;\\">
    <div class=\\"v-navigation-drawer__content\\">
      <div class=\\"v-list v-sheet theme--light v-list--dense v-list--nav\\">
        <div class=\\"v-subheader mt-2 theme--light\\"><span class=\\"nav-drawer-header\\">Table of Contents</span></div>
      </div> <a tabindex=\\"0\\" class=\\"link-text v-list-item v-list-item--link theme--light primary--text\\">
        <div class=\\"v-list-item__content\\">
          <div class=\\"v-list-item__title\\">title1</div>
        </div>
      </a><a tabindex=\\"0\\" class=\\"link-text v-list-item v-list-item--link theme--light primary--text\\">
        <div class=\\"v-list-item__content\\">
          <div class=\\"v-list-item__title\\">title2</div>
        </div>
      </a><a tabindex=\\"0\\" class=\\"link-text v-list-item v-list-item--link theme--light primary--text\\">
        <div class=\\"v-list-item__content\\">
          <div class=\\"v-list-item__title\\">title3</div>
        </div>
      </a>
    </div>
    <div class=\\"v-navigation-drawer__border\\"></div>
  </nav>
</div>"
`;
